provider/azure/environ.go

```diff
@@ -528,29 +532,54 @@ func (env *azureEnviron) StartInstance(ctx context.ProviderCallContext, args env
 		rootDisk = minRootDiskSize
 		args.Constraints.RootDisk = &rootDisk
 	}
-
-	// Identify the instance type and image to provision.
-	instanceSpec, err := findInstanceSpec(
-		ctx,
-		compute.VirtualMachineImagesClient{BaseClient: env.compute},
-		instanceTypes,
-		&instances.InstanceConstraint{
-			Region:      env.location,
-			Series:      args.InstanceConfig.Series,
-			Arches:      args.Tools.Arches(),
-			Constraints: args.Constraints,
-		},
-		imageStream,
-	)
-	if err != nil {
-		return nil, err
-	}
-	if rootDisk < instanceSpec.InstanceType.RootDisk {
-		// The InstanceType's RootDisk is set to the maximum
-		// OS disk size; override it with the user-specified
-		// or default root disk size.
-		instanceSpec.InstanceType.RootDisk = rootDisk
+	// Start the instance - if we get a quota error, that instance type is ignored
+	// and we'll try the next most expensive one, up to a reasonable number of attempts.
+tryagain:
+	for i := 0; i < 15; i++ {
+		// Identify the instance type and image to provision.
+		instanceSpec, err := findInstanceSpec(
+			ctx,
+			compute.VirtualMachineImagesClient{BaseClient: env.compute},
+			instanceTypes,
+			&instances.InstanceConstraint{
+				Region:      env.location,
+				Series:      args.InstanceConfig.Series,
+				Arches:      args.Tools.Arches(),
+				Constraints: args.Constraints,
+			},
+			imageStream,
+		)
+		if err != nil {
+			return nil, err
+		}
+		if rootDisk < instanceSpec.InstanceType.RootDisk {
+			// The InstanceType's RootDisk is set to the maximum
+			// OS disk size; override it with the user-specified
+			// or default root disk size.
+			instanceSpec.InstanceType.RootDisk = rootDisk
+		}
+		result, err := env.startInstance(ctx, args, instanceSpec, envTags)
+		if err == nil {
+			return result, nil
+		}
+		serviceErr, ok := errorutils.ServiceError(err)
+		if ok {
+			for _, d := range serviceErr.Details {
+				if code, ok := d["code"].(string); ok && code == "QuotaExceeded" {
```

Comment:
It'd be better to wrap these lines into a `func IsQuotaExceeded(err error)` in `./errorutils` package

Reply:
yeah, makes it nicer, ty